python/instrumentation/openinference-instrumentation-boto/src/openinference/instrumentation/boto/__init__.py

```diff
@@ -0,0 +1,169 @@
+import io
+import json
+from functools import wraps
+from importlib import import_module
+from inspect import signature
+from typing import IO, Any, Callable, Collection, Dict, Optional, Tuple, TypeVar, cast
+
+from botocore.client import BaseClient
+from botocore.response import StreamingBody
+from openinference.semconv.trace import MessageAttributes, SpanAttributes
+from opentelemetry import context as context_api
+from opentelemetry import trace as trace_api
+from opentelemetry.context import _SUPPRESS_INSTRUMENTATION_KEY
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor  # type: ignore
+from opentelemetry.trace import Tracer
+from opentelemetry.util.types import AttributeValue
+from wrapt import wrap_function_wrapper
+
+CallableType = TypeVar("CallableType", bound=Callable[..., Any])
+
+_MODULE = "botocore.client"
+__version__ = "0.1.0"
+_instruments = ("boto3 >= 1.28.57",)
+
+
+def _set_span_attribute(span: trace_api.Span, name: str, value: AttributeValue) -> None:
+    if value is not None:
+        if value != "":
+            span.set_attribute(name, value)
+    return
+
+
+class InstrumentedClient(BaseClient):  # type: ignore
+    """
+    Proxy class representing an instrumented boto client.
+    """
+
+    invoke_model: Callable[..., Any]
+    _unwrapped_invoke_model: Callable[..., Any]
+
+
+class BufferedStreamingBody(StreamingBody):  # type: ignore
+    _raw_stream: IO[bytes]
+
+    def __init__(self, raw_stream: IO[bytes], content_length: int) -> None:
+        super().__init__(raw_stream, content_length)
+        self._buffer: Optional[io.IOBase] = None
+
+    def read(self, amt: Optional[int] = None) -> bytes:
+        if self._buffer is None:
+            self._buffer = io.BytesIO(self._raw_stream.read())
+
+        output: bytes = self._buffer.read(amt)
+        return output
+
+    def reset(self) -> None:
+        # Reset the buffer to enable reading the stream again
+        if self._buffer is not None:
+            self._buffer.seek(0)
+
+
+def _client_creation_wrapper(tracer: Tracer) -> Callable[[CallableType], CallableType]:
+    def _client_wrapper(
+        wrapped: CallableType,
+        instance: Optional[Any],
+        args: Tuple[Any, ...],
+        kwargs: Dict[str, Any],
+    ) -> BaseClient:
+        @wraps(wrapped)
+        def create_instrumented_client(*args: Any, **kwargs: Any) -> BaseClient:
+            """Instruments and calls every function defined in TO_WRAP."""
```

Comment:
What is TO_WRAP?

Reply:
a docstring placeholder I forgot to fix 🥲